### PR TITLE
encryption: add test for temporary KMS service unavailable

### PR DESCRIPTION
**Pull request by hhwyt**
ref tikv#17410

This commit adds a new test to verify TiKV's recovery after temporary unavailability of the KMS service.

### What is changed and how it works?

Issue Number: Ref #17410

What's Changed:

```commit-message
This commit adds a new test to verify TiKV's recovery after temporary unavailability of the KMS service.
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
no.
```

### CODE REVIEW

File: components/encryption/src/master_key/kms.rs [hunk 1]
```diff
@@ -219,27 +231,62 @@ mod fake {
...
         async fn decrypt_data_key(&self, _ciphertext: &EncryptedKey) -> Result<Vec<u8>> {
+            let fail = (|| {
+                fail_point!("kms_api_timeout_decrypt", |val| val
+                    .and_then(|x| x.parse::<bool>().ok())
+                    .unwrap_or(false));
+                false
+            })();
+            if fail {
+                return Err(CloudError::ApiTimeout(box_err!("api timeout")));
+            }
```
Comment (overvenus) [inline components/encryption/src/master_key/kms.rs:242]:
```suggestion
            (|| {
                let fail = fail_point!("kms_api_timeout_decrypt", |val| val
                    .and_then(|x| x.parse::<bool>().ok())
                    .unwrap_or(false));
                if fail {
                    Err(CloudError::ApiTimeout(box_err!("api timeout")))
                } else {
                    Ok(())
                }
            })()?;
```

File: components/encryption/src/master_key/kms.rs [hunk 2]
```diff
@@ -211,6 +213,16 @@ mod fake {
...
         async fn generate_data_key(&self) -> Result<DataKeyPair> {
+            let fail = (|| {
+                fail_point!("kms_api_timeout_encrypt", |val| val
+                    .and_then(|x| x.parse::<bool>().ok())
+                    .unwrap_or(false));
+                false
+            })();
+            if fail {
+                return Err(CloudError::ApiTimeout(box_err!("api timeout")));
+            }
```
Comment (overvenus) [inline components/encryption/src/master_key/kms.rs:224]: See above comments.

File: components/encryption/src/master_key/kms.rs (hunk 1 above)
Comment (hhwyt) [reply components/encryption/src/master_key/kms.rs]: I refactored the `fail_point` part to the following simplified version. Let me know if you have any thoughts or suggestions!

```rust
        fail_point!(fail_point_name, |val| {
            val.and_then(|x| x.parse::<bool>().ok())
                .filter(|&fail| fail)
                .map(|_| Err(CloudError::ApiTimeout(box_err!("api timeout"))))
                .unwrap_or(Ok(()))
        });
```

**Review by LykxSassinator — APPROVED** [review]
LGTM

**Review by overvenus — APPROVED** [review]

**Pull request merged**